---
title: "insecta fasta"
output:
  html_notebook:
    code_folding: show
    df_print: paged
    fig_width: 12
    fig_height: 8
    fig_caption: yes
    toc: yes
    toc_depth: 6
    toc_float:
      collapsed: no
---

```{css, echo=FALSE}
.tocify .tocify-header {
  #position: fixed;
  #top: 50px;
  #left: 50px;
  width: 350px;
  #height: 400px;
  }
```

```{r setup}
suppressPackageStartupMessages({
  library(bold)
  library(taxize)
  library(dplyr)
  library(Biostrings)
  library(tidyverse)
})
```

```{r}
# phylum_name	class_name	order_name	family_name	genus_name	species_name	sequenceID	nucleotides
# otherArth_list

# List objects
#
# otherInsects_list

# Col_Carabidae_list
# Col_Chrysomelidae_list
# Col_Curculionidae_list
# Col_Staphylinidae_list
# Col_allothers_list
# 
# Dip_Sciaridae_list
# Dip_Cecidomyiidae_list
# Dip_Chironomidae_list
# Dip_Phoridae_list
# Dip_Ceratopogonidae_list
# Dip_allothers_list
# 
# Hym_Braconidae_list
# Hym_Formicidae_list
# Hym_Ichneumonidae_list
# Hym_Platygastridae_list
# Hym_allothers_list
# 
# Lep_Noctuidae_list
# Lep_Erebidae_list
# Lep_Sphingidae_list
# Lep_Geometridae_list
# Lep_Crambidae_list
# Lep_Gelechiidae_list
# Lep_Saturniidae_list
# Lep_allothers_list


# c(Col_list,diptera_list, Lep_list, Hym_list, otherArth_list) not included

ins.list = list(otherInsects_list, Col_Carabidae_list, Col_Chrysomelidae_list, Col_Curculionidae_list, Col_Staphylinidae_list,
                Col_allothers_list, Dip_Sciaridae_list, Dip_Cecidomyiidae_list, Dip_Chironomidae_list, Dip_Phoridae_list,
                Dip_Ceratopogonidae_list, Dip_allothers_list, Hym_Braconidae_list, Hym_Formicidae_list, Hym_Ichneumonidae_list,
                Hym_Platygastridae_list, Hym_allothers_list, Lep_Noctuidae_list, Lep_Erebidae_list, Lep_Sphingidae_list,
                Lep_Geometridae_list, Lep_Crambidae_list, Lep_Gelechiidae_list, Lep_Saturniidae_list, Lep_allothers_list)

saveRDS(ins.list, "./rds/ins.list.rds") # for backup
```

# tsvdf
```{r}
# Define tsvdf function to convert the ins.list to dataframe
tsvdf = function(bold_seqspec_list){
  temp.df.rbind =c()
  
  for(x in 1:length(bold_seqspec_list)){
    a = bold_seqspec_list[[x]]
    
    temp.df = data.frame("phylum_name" = a$name$phylum_name,   # V1
                         "class_name" = a$name$class_name,     # V2
                         "order_name" = a$name$order_name,     # V3
                         "family_name" = a$name$family_name,   # V4
                         "genus_name" = a$name$genus_name,     # V5
                         "species_name" = a$name$species_name, # V6
                         "sequenceID" = a$name$sequenceID,     # V7
                         "nucleotides" = a$name$nucleotides,   # V8
                         "markercode" = a$name$markercode,     # V9 added markercode
                         "country" = a$name$country)           # v10 added country
    
    temp.df.rbind = rbind(temp.df.rbind, temp.df)
    print(x)
  }
  return(temp.df.rbind)
}
ins.df = tsvdf(ins.list)
```

# select COI-5P
```{r}
head(ins.df)
unique(ins.df$markercode)
length(which(ins.df$markercode == "COI-5P"))

ins.df2 = filter(ins.df, markercode %in% c("COI-5P")) # filter only COI-5P
head(ins.df2)
```

# raw entries count COI-5P
```{r}
# glo
ins.df %>% filter(markercode %in% c("COI-5P")) %>% dim()
# reg
ins.df %>% filter(markercode %in% c("COI-5P")) %>% filter(country %in% temp_regional) %>% dim()
# loc
ins.df %>% filter(markercode %in% c("COI-5P")) %>% filter(country == "Malaysia") %>% dim()
```


# cleaning
```{r}
head(ins.df2)
str(ins.df2)

# Define tsv.clean function
tsv.clean = function(df){
  temp = df
  for (x in 8:2) {
    temp = temp[temp[,x] != "" ,]
  }
  print("removing empty taxa 8 to 2 done")
  
  temp[,6] = gsub(",", "_", temp[,6])
  print ("removing comma on V6 done")
  
  temp[,6] = gsub(" ", "_", temp[,6])
  print ("replacing space with underscore on V6 done")
  
  temp[,6] = sub("_sp.*", "_sp", temp[,6])
  print("removing addition species names after sp. done")
  
  #removing the dashes that appear to be the result of sequence alignment before going into the BOLD db
  temp[,8] = gsub("-", "", temp[,8])
  print("removing the dashes because alignment on col 8 done")
  
  temp$taxa = paste0(temp[,7], ";tax=p:", temp[,1], ",c:", temp[,2], ",o:", temp[,3],
                     ",f:", temp[,4], ",g:", temp[,5], ",s:", temp[,6],  ";")
  print("adding complet taxonomy on col 9 done")
  
  return(temp)
}

ins.df2.clean = tsv.clean(ins.df2)
```

# check empty info using fempty
```{r}
f.empty = function(cleandf){ # function to find empty entries for each column
  for(x in c(1:7,10)){
    a = colnames(cleandf)[x]
    print(a)
    b = length(which(cleandf[,x] == ""))
    print(b)
  }
}

f.empty(ins.df2.clean)
```

# check NAs using findNA
```{r}
findNA = function(df){
  d = c()
  for(x in rev(seq(1, 11, by = 1))){
    a = names(df)[x]
    b = sum(is.na(df[,x]))
    c = data.frame("column" = a, "nos.na"=b)
    d = rbind(d,c)
  }
  return(d)
  print(d)
}

findNA(ins.df2.clean)
```

# save to fasta files
```{r}
# global
db.global.fasta = DNAStringSet(ins.df2.clean$nucleotides)
names(db.global.fasta) = ins.df2.clean$taxa
db.global.fasta
range(db.global.fasta@ranges@width)
writeXStringSet(db.global.fasta, "./results/db.global2.fa")

# regional
db.regional = subset(ins.df2.clean, country %in% temp_regional)
db.regional.fasta = DNAStringSet(db.regional$nucleotides)
names(db.regional.fasta) = db.regional$taxa
db.regional.fasta
range(db.regional.fasta@ranges@width)
writeXStringSet(db.regional.fasta, "./results/db.regional2.fa")

# local
db.local = subset(ins.df2.clean, country == "Malaysia")
db.local.fasta = DNAStringSet(db.local$nucleotides)
names(db.local.fasta) = db.local$taxa
db.local.fasta
range(db.local.fasta@ranges@width)
writeXStringSet(db.local.fasta, "./results/db.local2.fa")
```

# save rdata
```{r}
gc()
save.image("./insecta.RData")
```


# Session info
```{r}
sessioninfo::session_info(pkgs="attached")
```